packages/eslint-plugin/docs/rules/no-extraneous-return-types.mdx

```diff
@@ -0,0 +1,67 @@
+---
+description: 'Disallow extraneous union types in return type annotations.'
+---
+
+import Tabs from '@theme/Tabs';
+import TabItem from '@theme/TabItem';
+
+> 🛑 This file is source code, not the primary documentation location! 🛑
+>
+> See **https://typescript-eslint.io/rules/no-extraneous-return-types** for documentation.
+
+Explicit return type annotations are optional in TypeScript. However, explicit return types do make it visually more clear what type is returned by a function.
+
+A function can include return type annotations that are never used. This can happen when refactoring code or by mistake. Having unused return types adds unnecessary complexity because it needs to be handled by code that calls the function.
+
+This rule reports cases where a function has an unused return type annotation.
+
+## Examples
+
+<Tabs>
+<TabItem value="❌ Incorrect">
+
+```ts
+function test1(): string | null {
+  return 'string';
+}
+
+function test2(): Promise<string | null> {
+  return Promise.resolve('string');
+}
+
+class A {
+  test(): string | null {
+    return 'string';
+  }
+}
+```
+
+</TabItem>
+<TabItem value="✅ Correct">
+
+```ts
+function test1(): string {
+  return 'string';
+}
+
+function test2(): Promise<string> {
+  return Promise.resolve('string');
+}
+
+class A {
+  test(): string {
+    return 'string';
+  }
+}
+```
+
+</TabItem>
+</Tabs>
+
+## When Not To Use It
+
+This rule can be completely redundant for projects without explicit return type annotations on functions. It works best when coupled with other rules that enforce explicit return type annotations such as [`explicit-function-return-type`](https://typescript-eslint.io/rules/explicit-function-return-type) or [`explicit-module-boundary-types`](https://typescript-eslint.io/rules/explicit-module-boundary-types).
```

Comment:
[Docs] I think this accidentally implies that rule is _only_ useful with those functions. Which is not true, I've found myself occasionally wanting it in projects that don't use those rules or isolatedDeclarations. Even though I hate unnecessary types I still sometimes use them to make functions more clear or get better return types.

I think this should instead note the actual edge cases that make the rule less useful. Which are ... not that many that I can think of. Maybe:

* If the functions are early stage and will likely be evolved/widened later
* If you're explicitly trying to match some external, more-permissive interface

...along with the standard notes about considering using inline disables in those cases.

Reply:
Yes, I agree.

The current explanation of `This rule can be completely redundant for projects...` is just wrong (and may make incoming developers miss a lot of the value of such a rule). I usually omit return type annotation in projects, and I still have dozens of functions I choose to annotate.

I updated this to match this, let me know what you think 👍